thread_sync.c

```diff
@@ -38,25 +71,41 @@ struct queue_sleep_arg {
 
 #define MUTEX_ALLOW_TRAP FL_USER1
 
+static rb_mutex_t * mutex_ptr(VALUE obj);
+static bool mutex_has_live_reservation_p(rb_mutex_t *mutex);
+static bool mutex_reservation_can_have_p(rb_mutex_t *mutex, rb_thread_t *for_thread);
+/* needs to be VALUE, not rb_mutex_t, so it can update the reservation through the write-barrier */
+static bool mutex_reserve_if_available(VALUE mutex_value, rb_thread_t *for_thread);
+
 static void
 sync_wakeup(struct ccan_list_head *head, long max)
 {
     struct sync_waiter *cur = 0, *next;
+    long woken = 0;
 
     ccan_list_for_each_safe(head, cur, next, node) {
-        ccan_list_del_init(&cur->node);
-
         if (cur->th->status != THREAD_KILLED) {
             if (cur->th->scheduler != Qnil && cur->fiber) {
                 rb_fiber_scheduler_unblock(cur->th->scheduler, cur->self, rb_fiberptr_self(cur->fiber));
```

Comment:
I wasn't sure if I should implement this for the fiber-scheduler case as well - I was half-wondering if this kind of fix should actually be applied in the fiber scheduler implementaiton(s)? 

I think I can make it work in here though if that's preferable.